src/main/java/sirius/web/controller/ControllerDispatcher.java

```diff
@@ -71,42 +79,60 @@ public int getPriority() {
+                InputStreamHandler ish = new InputStreamHandler();
+                ctx.setContentHandler(ish);
 
-    @Override
-    public boolean dispatch(WebContext ctx) throws Exception {
-        return route(ctx, false);
+                return newCtx -> preparePerformRoute(newCtx, route, params, ish);
+            }
+        }
+
+        return null;
     }
 
-    private boolean route(final WebContext ctx, boolean preDispatch) {
+    private String determineEffectiveURI(WebContext ctx) {
         String uri = ctx.getRawRequestedURI();
         if (uri.endsWith("/") && !"/".equals(uri)) {
             uri = uri.substring(0, uri.length() - 1);
         }
-        for (final Route route : getRoutes()) {
-            if (tryExecuteRoute(ctx, preDispatch, uri, route)) {
-                return true;
-            }
-        }
-        return false;
+        return uri;
     }
 
     @SuppressWarnings("squid:S1698")
-    @Explain("Strings have been intered for performance reasons.")
-    private boolean tryExecuteRoute(WebContext ctx, boolean preDispatch, String uri, Route route) {
-        try {
-            final List<Object> params = route.matches(ctx, uri, preDispatch);
-            if (params == Route.NO_MATCH) {
-                // Route did not match...
-                return false;
+    @Explain("We acutally can use object identity here as this is a marker object.")
+    private List<Object> shouldExecute(WebContext ctx, String uri, Route route, boolean preDispatch) {
+        final List<Object> params = route.matches(ctx, uri, preDispatch);
+        if (params == Route.NO_MATCH) {
+            // Route did not match...
+            return params;
+        }
+        // Check if interceptors permit execution of route...
+        for (Interceptor interceptor : interceptors) {
+            if (!interceptor.shouldExecuteRoute(ctx, route.isJSONCall(), route.getController())) {
+                return Route.NO_MATCH;
             }
+        }
 
-            // Check if interceptors permit execution of route...
-            for (Interceptor interceptor : interceptors) {
-                if (!interceptor.shouldExecuteRoute(ctx, route.isJSONCall(), route.getController())) {
-                    return false;
+        return params;
+    }
+
+    private void preparePerformRoute(WebContext ctx,
+                                     Route route,
+                                     List<Object> params,
+                                     InputStreamHandler inputStreamHandler) {
+        try {
+            if (firewall != null && !route.getMethod().isAnnotationPresent(Unlimited.class)) {
+                if (firewall.handleRateLimiting(ctx,
+                                                Optional.ofNullable(route.getMethod().getAnnotation(Limited.class))
+                                                        .map(Limited::value)
+                                                        .orElse(Limited.HTTP))) {
+                    return;
```

Comment:
HTTP Status Code 429 "Too many requests" statt einfach ins leere laufen zu lassen?

Reply:
If handle returns true, we expect the firewall to have the request handled (e.g. with said status code)